hack/preload-images/preload_images.go

```diff
@@ -129,11 +129,26 @@ func collectK8sVers() ([]string, error) {
 		}
 		k8sVersions = recent
 	}
-	return append([]string{
+	versions := append([]string{
 		constants.DefaultKubernetesVersion,
 		constants.NewestKubernetesVersion,
 		constants.OldestKubernetesVersion,
-	}, k8sVersions...), nil
+	}, k8sVersions...)
+	return removeDuplicates(versions), nil
+}
+
+func removeDuplicates(versions []string) []string {
```

Comment:
This seems like a reusable function that we might want to keep in a shared library (although I have not yet checked elsewhere in the code for whether we de duplicate string slices).

Also, would it be possible to add a test?

I think the `range` function is little easier to read, here is what I came up with:

```
// dedupeStrings deduplicates a string slice, removes empty strings.
func dedupeStrings(dupes []string) []string {
	var uniques []string
	m := make(map[string]struct{}, len(dupes))
	for _, v := range dupes {
		if _, ok := m[v]; !ok {
			m[v] = struct{}{}
			uniques = append(uniques, v)
		}
	}
	return uniques
}
```

Test

```
func TestDedupeStrings(t *testing.T) {
	testCases := []struct {
		desc  string
		slice []string
		want  []string
	}{
		{
			desc:  "no duplicates",
			slice: []string{"alpha", "bravo", "charlie"},
			want:  []string{"alpha", "bravo", "charlie"},
		},
		{
			desc:  "adjacent duplicates",
			slice: []string{"alpha", "bravo", "bravo", "charlie"},
			want:  []string{"alpha", "bravo", "charlie"},
		},
		{
			desc:  "non-adjacent duplicates",
			slice: []string{"alpha", "bravo", "alpha", "charlie"},
			want:  []string{"alpha", "bravo", "charlie"},
		},
		{
			desc:  "multiple duplicates",
			slice: []string{"alpha", "bravo", "alpha", "alpha", "charlie", "charlie", "alpha", "bravo"},
			want:  []string{"alpha", "bravo", "charlie"},
		},
		{
			desc:  "unsorted duplicates",
			slice: []string{"charlie", "bravo", "alpha", "bravo"},
			want:  []string{"charlie", "bravo", "alpha"},
		},
	}
	for _, tc := range testCases {
		t.Run(tc.desc, func(t *testing.T) {
			got := dedupeStrings(tc.slice)
			if diff := cmp.Diff(got, tc.want); diff != "" {
				t.Errorf("dedupeStrings(%v) = %v, want: %v", tc.slice, got, tc.want)
			}
		})
	}
}
```

Reply:
That makes sense, I was trying to go the path of removing duplicated elements from the original slice, but it's cleaner to go the path of creating a new slice and add unique elements.

I can also add the test, we don't currently run tests in the `hack` dir, but I can update our `make test` command to check for tests in there as well.